src/hotspot/share/opto/loopConditionalPropagation.cpp

```diff
@@ -0,0 +1,1213 @@
+           the CMove (corresponding if not yet eliminated). Inconsistencies between passes of this optimization can cause
+           issues so skip merging of CMove entirely.
+           */
+          continue;
+        }
+        const Type* t = find_type_between(n, in, dom);
+        // and check if the type was updated from other region inputs
+        uint k = 1;
+        for (; k < c->req(); k++) {
+          if (k == in_idx) {
+            continue;
+          }
+          Node* other_in = c->in(k);
+          const Type* type_at_in = find_type_between(n, other_in, dom);
+          if (type_at_in == nullptr) {
+            break;
+          }
+          t = t->meet_speculative(type_at_in);
+        }
+        // If that's the case, record type update
+        if (k == c->req()) {
+          const Type* current_type = find_prev_type_between(n, in, dom);
+
+          if (_iterations > 1) {
+            t = current_type->filter(t);
+            const Type* prev_t = t;
+            const Type* prev_round_t = nullptr;
+            if (_prev_updates != nullptr && _prev_updates->control() == c) {
+              prev_round_t = _prev_updates->type_if_present(n);
+            }
+            if (prev_round_t == nullptr) {
+              prev_round_t = current_type;
+            }
+            if (prev_round_t != nullptr) {
+              t = prev_round_t->filter(t);
+              assert(t == prev_t, "");
+              t = saturate(t, prev_round_t, nullptr);
+              if (c->is_Loop() && t != prev_round_t) {
+                extra = true;
+              }
+            }
+          }
+          t = current_type->filter(t);
+
+          if (t != current_type) {
+            assert(narrows_type(current_type, t), "");
+            if (record_update(c, n, current_type, t)) {
+              enqueue_uses(n, c);
+            }
+          }
+        }
+      }
+      updates = updates->prev();
+      assert(updates != nullptr || _dom_updates == nullptr || _phase->is_dominator(c, in) || C->has_irreducible_loop(), "");
+    }
+  } else if (c->is_IfProj()) {
+    Node* iff = c->in(0);
+    assert(iff->is_If(), "");
+    if (!(iff->is_CountedLoopEnd() && iff->as_CountedLoopEnd()->loopnode() != nullptr &&
+          iff->as_CountedLoopEnd()->loopnode()->is_strip_mined())) {
+      Node* bol = iff->in(1);
+      if (iff->is_OuterStripMinedLoopEnd()) {
+        assert(iff->in(0)->in(0)->in(0)->is_CountedLoopEnd(), "");
+        bol = iff->in(0)->in(0)->in(0)->in(1);
+      }
+      if (bol->Opcode() == Op_Opaque4) {
+        bol = bol->in(1);
+      }
+      if (bol->is_Bool()) {
+        Node* cmp = bol->in(1);
+        if (cmp->Opcode() == Op_CmpI || cmp->Opcode() == Op_CmpU ||
+            cmp->Opcode() == Op_CmpL || cmp->Opcode() == Op_CmpUL) {
+          Node* cmp1 = cmp->in(1);
+          Node* cmp2 = cmp->in(2);
+          sync(iff);
+          // narrowing the type of a LoadRange could cause a range check to optimize out and a Load to be hoisted above
```

Comment:
Not sure if you are still working on this one. Here you can remove the range check if you pin all the nodes that depend on this check. Similar logic can be found in range check smearing.

Reply:
Yes, I'm still working on this. I'll update this PR soon, hopefully. I reworked it quite a bit.

Reply:
Some more observations: When removing an `IfNode`, not only for `LoadNode`, you will need to pin all nodes that `depends_only_on_test` at that point (e.g. `ConstraintCast`), since the node does not only depend on the immediate dominating test (if any) but on the whole sequence of control nodes leading to that position. This can be done by e.g upgrading a `RegularDependency` `ConstraintCast` to one with `StrongDependency`. However, this action can lead to the node not able to float as freely. So I think before completing all loop opts, you should refrain from removing any `IfNode` of which the taken path has at least 1 node that `depends_only_on_test`. Pruning the untaken branch should still be done to simplify the graph. I think you have probably thought about this but just in case it can help.